### PR TITLE
CONTRIBUTING: Added documentation on how to add a new Provider

### LINKED ISSUE #248: Minimal documentation about how to add support for a cloud provider
### Abstract

Could you add documentation explaining how a given cloud vendor/terraform provider implementer can add support for their provider inside terracognita?
What are the difficult aspects? How is it tested? Documented?

On the code generation perspective, could you give a brief overview about how terracognita works from a software architecture perspective?

I'm available for review as an external contributor :)

### PR DESCRIPTION
**Pull request by xescugc**
We had documentation on how to add new Resources but not new ProvidersW

Closes #248

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -61,6 +61,39 @@ This basically builds `terracognita` with the current sources.
 You also need to install other code dependencies not mandatory in the runtime environment:
   * [enumer](https://github.com/dmarkham/enumer) is used to generate some code.
   * [goimports](https://godoc.org/golang.org/x/tools/cmd/goimports) is used to format / organize code and imports. CI will perform a check on this, we highly recommend to run `$ goimports -w <your-modified-files>`.
+  * [mockgen](https://github.com/golang/mock) is used to generates the mocks for testing
+
+#### Add a new Provider
+
+:warning: It has to support `github.com/hashicorp/terraform-plugin-sdk@v2.0.0` or higher, providers with `v1` will not work anymore :warning:
+
+All the provider names of packages and so one are the ones used by the provider itself, the one in which all the resources are prefixed with (ex: `aws_instance` is `aws`)
```
Comment (xlr-8) [inline CONTRIBUTING.md:70]: Couldn't that also be `registry.terraform.io/hashicorp/aws`, or that's only in cases of modules ?

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -61,6 +61,39 @@ This basically builds `terracognita` with the current sources.
...
+
+**Provider**
+
+All providers have to follow the [provider.Provider](https://github.com/cycloidio/terracognita/blob/master/provider/provider.go#L14) interface in order to be able to be integrated to the codebase. The best way would be to follow any of the other providers and modify it accordingly to your needs. Also reading the documentation of the interface will help a lot in any question regarding what means any of the functions.
```
Comment (xlr-8) [inline CONTRIBUTING.md:74]: any questions*

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -61,6 +61,39 @@ This basically builds `terracognita` with the current sources.
...
+
+Resources are each of one of the Terraform resources that this specific provider has and can import using TerraCognita. The main idea is that the `provider.Provider.Resources` returns all the instances of one resource. That information has to be fetched from the provider API using the sdk initialized before and for each instance we have to build the Terraform resource ID and return the resource. This then will be used by Terraform to read all the needed data from the provider to load the resource Schema (no need to do anything as this is abstracted into the caller of that function).
+
+To know what's the ID expected by Terraform you can check on the Terraform documentation of the resource ([doc](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/instance#import)[TODO: link docs and tf code], sometimes is more complex and it needs several elements to build an ID ([doc](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/iam_role_policy#import) and [code](https://github.com/hashicorp/terraform-provider-aws/blob/main/aws/resource_aws_iam_role_policy.go#L171)) so we have to set it on the right format for then be read for Terraform.
```
Comment (xlr-8) [inline CONTRIBUTING.md:84]: > [TODO: link docs and tf code]

No no

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -61,6 +61,39 @@ This basically builds `terracognita` with the current sources.
...
+**Provider**
+
+All providers have to follow the [provider.Provider](https://github.com/cycloidio/terracognita/blob/master/provider/provider.go#L14) interface in order to be able to be integrated to the codebase. The best way would be to follow any of the other providers and modify it accordingly to your needs. Also reading the documentation of the interface will help a lot in any question regarding what means any of the functions.
+
+New providers have to be on the root with the package name being the name of the provider on Terraform (ex: Amazon is `aws`)
+
+The `NewProvider` will need all the information required to initialize the specific provider client(sdk), and the specific provider configuration so then Terraform can use it to read information remotely and load it to the Schemas. All the information has to be sent via parameters, no ENV variables.
+
+**Resources**
+
```
Comment (xlr-8) [inline CONTRIBUTING.md:81]: I'd had a short notice such as: `In the following paragraphs, we will use AWS as example as it is the most advanced of the project, but the logic could apply similarly across the other providers too.`

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -61,6 +61,39 @@ This basically builds `terracognita` with the current sources.
...
+Resources are each of one of the Terraform resources that this specific provider has and can import using TerraCognita. The main idea is that the `provider.Provider.Resources` returns all the instances of one resource. That information has to be fetched from the provider API using the sdk initialized before and for each instance we have to build the Terraform resource ID and return the resource. This then will be used by Terraform to read all the needed data from the provider to load the resource Schema (no need to do anything as this is abstracted into the caller of that function).
+
+To know what's the ID expected by Terraform you can check on the Terraform documentation of the resource ([doc](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/instance#import)[TODO: link docs and tf code], sometimes is more complex and it needs several elements to build an ID ([doc](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/iam_role_policy#import) and [code](https://github.com/hashicorp/terraform-provider-aws/blob/main/aws/resource_aws_iam_role_policy.go#L171)) so we have to set it on the right format for then be read for Terraform.
+
```
Comment (xlr-8) [inline CONTRIBUTING.md:85]: I'd add a paragraph for the "multiple style resources" - eg. `security_group` and `security_group_rules` - which one to pick, how to decide, etc.

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -61,6 +61,39 @@ This basically builds `terracognita` with the current sources.
...
+New providers have to be on the root with the package name being the name of the provider on Terraform (ex: Amazon is `aws`)
+
+The `NewProvider` will need all the information required to initialize the specific provider client(sdk), and the specific provider configuration so then Terraform can use it to read information remotely and load it to the Schemas. All the information has to be sent via parameters, no ENV variables.
+
+**Resources**
+
+Resources are each of one of the Terraform resources that this specific provider has and can import using TerraCognita. The main idea is that the `provider.Provider.Resources` returns all the instances of one resource. That information has to be fetched from the provider API using the sdk initialized before and for each instance we have to build the Terraform resource ID and return the resource. This then will be used by Terraform to read all the needed data from the provider to load the resource Schema (no need to do anything as this is abstracted into the caller of that function).
```
Comment (xlr-8) [inline CONTRIBUTING.md:82]: I'd replaced all "sdk" by "SDK"

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -61,6 +61,39 @@ This basically builds `terracognita` with the current sources.
...
+
+As providers have a lot of resources we have created custom sdks on top of the provider sdks that we use to simplify things. Those wrappers are autogenerated from a configuration that is custom for each Provider. We do this so when we have to add new resources is easy, one of the main reasons is to also avoid pagination and always return all the resources directly. For example in `aws` we have the [configuration](https://github.com/cycloidio/terracognita/blob/master/aws/cmd/functions.go) which has the [go:generate](https://github.com/cycloidio/terracognita/blob/master/aws/reader/connector.go#L56) that generates the end [reader](https://github.com/cycloidio/terracognita/blob/master/aws/reader/reader.go). We have similar things for the other sdks
```
Comment (xlr-8) [inline CONTRIBUTING.md:86]: > We do this so when we have to add new resources is easy, one of the main reasons is to also avoid pagination and always return all the resources directly.

We do this to make adding new resources easy; by automating the process and always returning all elements - thus avoiding repetitive code writing and pagination logic.

> We have similar things for the other sdks

We have similar logic for the other SDKs*

File: CONTRIBUTING.md [hunk 8]
```diff
@@ -61,6 +61,39 @@ This basically builds `terracognita` with the current sources.
...
+
+The next point on the documentation explains how to add new resources to a provider which will help on doing it from 0 too.
```
Comment (xlr-8) [inline CONTRIBUTING.md:88]: > from 0 too.

from zero.

File: CONTRIBUTING.md [hunk 9]
```diff
@@ -61,6 +61,39 @@ This basically builds `terracognita` with the current sources.
...
+
+*Note:* If a provider does not have the `Import` function on that specific resource (meaning it cannot be imported by `terraform import ...` it'll not be imported. But we can dynamically add that function so it can be imported as on [this example](https://github.com/cycloidio/terracognita/blob/dae282137c905c90602ae7e0d82cc7601ff1fc85/aws/resources.go#L1629)
```
Comment (xlr-8) [inline CONTRIBUTING.md:90]: > as on this example

as in this example*

File: CONTRIBUTING.md [hunk 10]
```diff
@@ -61,6 +61,39 @@ This basically builds `terracognita` with the current sources.
...
+
+**CMD**
+
+On the `cmd/` pkg you'll need to add the provider subcommand which you can mainly check how the other providers are done but it requires the initialization of the provider and the call to the main function `provider.Import()` with all the required parameters.
+
+Aside from that you also need to add the `resources` subcommand which will list all the resources supported by the provider in TerraCognita
```
Comment (xlr-8) [inline CONTRIBUTING.md:97]: Perhaps add a section for documentation: where to add that a new provider is supported, the version, the list of supported resources, etc.

**Review by xlr-8 — CHANGES_REQUESTED** [review]
Nice! Couple comments/typos, the overall looks :ok_hand:

File: CONTRIBUTING.md (hunk 1 above)
Comment (xescugc) [reply CONTRIBUTING.md:70]: This one in particualar is the `provider.Provider.Source()`, the name is basically the resource prefix, so `aws`.

File: CONTRIBUTING.md (hunk 10 above)
Comment (xescugc) [reply CONTRIBUTING.md:97]: Yep you are right

**Comment by xescugc** [comment]
Done

File: CONTRIBUTING.md [hunk 11]
```diff
@@ -71,31 +71,39 @@ All the provider names of packages and so one are the ones used by the provider
...
+To know what's the ID expected by Terraform you can check on the Terraform documentation of the resource ([doc](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/instance#import), sometimes is more complex and it needs several elements to build an ID ([doc](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/iam_role_policy#import) and [code](https://github.com/hashicorp/terraform-provider-aws/blob/main/aws/resource_aws_iam_role_policy.go#L171)) so we have to set it on the right format for then be read for Terraform.
 
-The next point on the documentation explains how to add new resources to a provider which will help on doing it from 0 too.
+There are some specific cases in which a resource is imported with another one, on the AWS case that would be `aws_security_groups`, so some times one resource can generate also other resources when then it's pulled by Terraform.
```
Comment (xlr-8) [inline CONTRIBUTING.md:88]: It was more about the "choice" to be made. Either `security_group` with rules defined inline, or `security_group` + `security_group_rules` associated

**Review by xlr-8 — COMMENTED** [review]
:+1: nice! One remaining comment

**Pull request merged**